src/shared/install.c

```diff
@@ -3771,7 +3800,12 @@ int unit_file_get_list(
                             !strv_contains(states, unit_file_state_to_string(f->state)))
                                 continue;
 
-                        r = hashmap_put(h, basename(f->path), f);
+                        _cleanup_free_ char *fname = NULL;
+                        r = path_extract_filename(f->path, &fname);
+                        if (r < 0)
+                                return r;
+
+                        r = hashmap_put(h, fname, f);
```

Comment:
IIUC the CI errors are caused by this line - you put a pointer to a key (`fname`) to the hashmap, but once `fname` here goes out of scope the memory for that pointer gets freed and subsequent hashmap_get() calls will use invalid memory when comparing hashmap keys.

Using something like:
```suggestion
                        r = hashmap_put(h, TAKE_PTR(fname), f);
```
should fix that.

This will then make a couple of tests complain about memory leaks, since the keys are now on the heap. Tweaking their cleanup handlers should take care of that:
```diff
diff --git a/src/test/test-install-root.c b/src/test/test-install-root.c
index c55445079c..f12e765325 100644
--- a/src/test/test-install-root.c
+++ b/src/test/test-install-root.c
@@ -591,7 +591,7 @@ TEST(preset_and_list) {
         UnitFileState state;
         bool got_yes = false, got_no = false;
         UnitFileList *fl;
-        _cleanup_hashmap_free_ Hashmap *h = NULL;
+        _cleanup_(hashmap_free_free_keyp) Hashmap *h = NULL;
 
         CLEANUP_ARRAY(changes, n_changes, install_changes_free);
 
diff --git a/src/test/test-load-fragment.c b/src/test/test-load-fragment.c
index 8d2cec0c01..bf6c5a16ee 100644
--- a/src/test/test-load-fragment.c
+++ b/src/test/test-load-fragment.c
@@ -44,7 +44,7 @@ _unused_ ConfigPerfItemLookup unused_lookup = load_fragment_gperf_lookup;
 
 TEST_RET(unit_file_get_set) {
         int r;
-        _cleanup_hashmap_free_ Hashmap *h = NULL;
+        _cleanup_(hashmap_free_free_keyp) Hashmap *h = NULL;
         UnitFileList *p;
 
         h = hashmap_new(&unit_file_list_hash_ops_free);
```

Reply:
Hmm, I think it's better to avoid touching the functions where hashmap is mixed in the play. It's better to systematically change those to `hashmap_ensure_put` in the future.

Reply:
@mrc0mmand Ah okay, thank you. I understand but I would've never had know that `hashmap_put`  was the issue. I didn't find anything that pointed to that being the issue. How were you able to determine it was that line causing a use-after-free? I want to be able to avoid such issues later in the future.

Reply:
> Hmm, I think it's better to avoid touching the functions where hashmap is mixed in the play. They need to be systematically changed to `hashmap_ensure_put` in the future.

I'll revert the changes I did at
```suggestion
                        r = hashmap_put(h, basename(f->path), f);
```

and add a comment  referring to your suggestion about not touching functions dealing with `hashmap` in the case of removing `basename()` instance

Reply:
> @mrc0mmand Ah okay, thank you. I understand but I would've never had know that `hashmap_put` was the issue. I didn't find anything that pointed to that being the issue. How were you able to determine it was that line causing a use-after-free? I want to be able to avoid such issues later in the future.

The CI jobs that run with sanitizers failed with this error:

```
==11132==ERROR: AddressSanitizer: heap-use-after-free on address 0x502000001810 at pc 0x73a788921c00 bp 0x7fff771ab050 sp 0x7fff771aa810
READ of size 3 at 0x502000001810 thread T0
    #0 0x73a788921bff  (/usr/lib/clang/17/lib/linux/libclang_rt.asan-x86_64.so+0x121bff) (BuildId: beb0a35f2d0823fd6420f355da05e49e71d737f5)
    #1 0x73a7871a1151 in base_bucket_scan /systemd-meson-build/../build/src/basic/hashmap.c:1228:29
    #2 0x73a7871a2813 in _hashmap_get /systemd-meson-build/../build/src/basic/hashmap.c:1360:15
    #3 0x73a786f13a57 in hashmap_get /systemd-meson-build/../build/src/basic/hashmap.h:170:16
    #4 0x73a786f13a57 in unit_file_get_list /systemd-meson-build/../build/src/shared/install.c:3781:29
    #5 0x5cc580b16aef in test_unit_file_get_set /systemd-meson-build/../build/src/test/test-load-fragment.c:53:13
    #6 0x5cc580b34a93 in run_test_table /systemd-meson-build/../build/src/shared/tests.h:162:42
    #7 0x5cc580b34a93 in main /systemd-meson-build/../build/src/test/test-load-fragment.c:1105:1
    #8 0x73a786643ccf  (/usr/lib/libc.so.6+0x25ccf) (BuildId: c0caa0b7709d3369ee575fcd7d7d0b0fc48733af)
    #9 0x73a786643d89 in __libc_start_main (/usr/lib/libc.so.6+0x25d89) (BuildId: c0caa0b7709d3369ee575fcd7d7d0b0fc48733af)
    #10 0x5cc580b16614 in _start (/systemd-meson-build/test-load-fragment+0x1a614) (BuildId: f5462c4fc2c6e045ec3592cb80f369329d7be832)

0x502000001810 is located 0 bytes inside of 16-byte region [0x502000001810,0x502000001820)
freed by thread T0 here:
    #0 0x73a78895a952  (/usr/lib/clang/17/lib/linux/libclang_rt.asan-x86_64.so+0x15a952) (BuildId: beb0a35f2d0823fd6420f355da05e49e71d737f5)
    #1 0x73a786f13d74 in freep /systemd-meson-build/../build/src/basic/alloc-util.h:106:22
    #2 0x73a786f13d74 in unit_file_get_list /systemd-meson-build/../build/src/shared/install.c:3813:17
    #3 0x5cc580b16aef in test_unit_file_get_set /systemd-meson-build/../build/src/test/test-load-fragment.c:53:13
    #4 0x5cc580b34a93 in run_test_table /systemd-meson-build/../build/src/shared/tests.h:162:42
    #5 0x5cc580b34a93 in main /systemd-meson-build/../build/src/test/test-load-fragment.c:1105:1
    #6 0x73a786643ccf  (/usr/lib/libc.so.6+0x25ccf) (BuildId: c0caa0b7709d3369ee575fcd7d7d0b0fc48733af)

previously allocated by thread T0 here:
    #0 0x73a78895b989 in malloc (/usr/lib/clang/17/lib/linux/libclang_rt.asan-x86_64.so+0x15b989) (BuildId: beb0a35f2d0823fd6420f355da05e49e71d737f5)
    #1 0x73a7888be986  (/usr/lib/clang/17/lib/linux/libclang_rt.asan-x86_64.`so+0xbe986) (BuildId: beb0a35f2d0823fd6420f355da05e49e71d737f5)
    #2 0x73a7871ff1aa in path_extract_filename /systemd-meson-build/../build/src/basic/path-util.c:1088:13
    #3 0x73a786f13d12 in unit_file_get_list /systemd-meson-build/../build/src/shared/install.c:3804:29
    #4 0x5cc580b16aef in test_unit_file_get_set /systemd-meson-build/../build/src/test/test-load-fragment.c:53:13
    #5 0x5cc580b34a93 in run_test_table /systemd-meson-build/../build/src/shared/tests.h:162:42
    #6 0x5cc580b34a93 in main /systemd-meson-build/../build/src/test/test-load-fragment.c:1105:1
    #7 0x73a786643ccf  (/usr/lib/libc.so.6+0x25ccf) (BuildId: c0caa0b7709d3369ee575fcd7d7d0b0fc48733af)

SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib/clang/17/lib/linux/libclang_rt.asan-x86_64.so+0x121bff) (BuildId: beb0a35f2d0823fd6420f355da05e49e71d737f5) 
```

In the first part of that error you can see that the call-chain ... -> `hashmap_get()` -> `_hashmap_get()` -> `base_bucket_scan()` leads to a heap-use after free. The second part of that error tells you what freed that accessed value, and the last part tells you what allocated that value in the first place, which was `path_extract_filename()` at line 3804 in `src/shared/install.c`. So from there you can tell that the value you parsed using `path_extract_filename()` and pushed to the hashmap (somehow, in the same function) was deallocated prematurely, and later access to the hashmap tried to read that freed value.